src/psyclone/psyir/backend/fortran.py

```diff
@@ -601,34 +601,54 @@ def gen_decls(self, symbol_table, args_allowed=True):
-                    for symbol in symbol_table.symbols
-                    if isinstance(symbol, RoutineSymbol)]):
-            raise VisitorError(
-                "Routine symbols without a global or local interface are not "
-                "supported by the Fortran back-end.")
+        # Keep a record of whether we've already checked for any wildcard
+        # imports to save doing so repeatedly
+        wildcard_imports_checked = False
+        first_wildcard_import = None
+
+        routine_symbols = [symbol for symbol in symbol_table.symbols
+                           if isinstance(symbol, RoutineSymbol)]
+        for sym in routine_symbols:
+            if (not isinstance(sym.interface, (LocalInterface,
+                                               GlobalInterface)) and
+                    sym.name.upper() not in FORTRAN_INTRINSICS):
+                if not wildcard_imports_checked:
+                    first_wildcard_import = \
+                        symbol_table.get_first_wildcard_import()
+                    wildcard_imports_checked = True
+                if not first_wildcard_import:
+                    raise VisitorError(
+                        "Routine symbol '{0}' does not have a global or local"
+                        "interface, is not a Fortran intrinsic and there is no"
+                        " wildcard import which could bring it into scope. "
+                        "This is not supported by the Fortran back-end.".
+                        format(sym.name))
 
         # Does the symbol table contain any symbols with a deferred
-        # interface (i.e. we don't know how they are brought into scope) that
-        # are not KIND parameters?
-        unresolved_datasymbols = symbol_table.get_unresolved_datasymbols(
-            ignore_precision=True)
+        # interface (i.e. we don't know how they are brought into scope)
+        unresolved_datasymbols = symbol_table.get_unresolved_datasymbols()
+
         if unresolved_datasymbols:
-            symbols_txt = ", ".join(
-                ["'" + sym + "'" for sym in unresolved_datasymbols])
-            raise VisitorError(
-                "The following symbols are not explicitly declared or imported"
-                " from a module (in the local scope) and are not KIND "
-                "parameters: {0}".format(symbols_txt))
+            # We do have unresolved symbols. Is there at least one wildcard
+            # import which could be bringing them into scope?
+            if not wildcard_imports_checked:
+                first_wildcard_import = \
+                    symbol_table.get_first_wildcard_import()
+            if not first_wildcard_import:
+                symbols_txt = ", ".join(
+                    ["'" + sym + "'" for sym in unresolved_datasymbols])
+                raise VisitorError(
+                    "The following symbols are not explicitly declared or "
+                    "imported from a module (in the local scope) and there "
```

Comment:
Maybe I am missing something but why the "(in the local scope)"? Isn't get_first/has_wildcard_import checks for wildcard imports at any level?

Reply:
No, it was me getting a bit confused. Fixed now.